src/porepy/numerics/ad/_ad_utils.py

```diff
@@ -623,4 +622,19 @@ def parse(self, mdg: pp.MixedDimensionalGrid) -> sps.spmatrix:
 
         else:
             # This is a standard discretization; wrap it in a diagonal sparse matrix.
-            return sps.block_diag(mat, format="csr")
+
+            if all([m.format == "dia" for m in mat]):
+                # If all matrices are of dia-format, we can try to use a special method
+                # for forming a sparse dia matrix from the blocks. This is more
+                # efficient than the below csr-based method. However, the matrices can
+                # only be nonzero along their main diagonal. This condition does not
+                # hold true for all dia-matrices, so there is a chance a ValueError may
+                # be raised. If so, we let it pass and fall back to the csr-based
+                # method.
+                try:
+                    return pp.matrix_operations.sparse_dia_from_sparse_blocks(mat)
+                except ValueError:
+                    # Use the csr-based method below.
+                    pass
+
+            return pp.matrix_operations.csr_matrix_from_sparse_blocks(mat)
```

Comment:
This is one of the format checks performed in sparse_dia_from_sparse_blocks. Do you suggest to add one more layer of safety checks to circumvent unnecessary ValueErrors?

Reply:
No, the other calls pass matrices that are known to be diagonal, and it is a very controlled environment. In this particular case, `mat` contains matrices obtained by some discretization method, hence the prudence.